File: BuddyGrammarKit/Sources/BuddyGrammarKit/ReviewableCorrectionProposal.swift
import Foundation

public enum BuddyRewriteIntent: String, CaseIterable, Identifiable, Sendable {
    case fix
    case shorten
    case clearer
    case friendly
    case formal

    public var id: String { rawValue }

    public var title: String {
        switch self {
        case .fix: "Fix"
        case .shorten: "Shorten"
        case .clearer: "Clearer"
        case .friendly: "Friendly"
        case .formal: "Formal"
        }
    }

    public func instruction(appendingTo base: String) -> String {
        let transformation = switch self {
        case .fix:
            "Fix only clear grammar, spelling, punctuation, and capitalization errors."
        case .shorten:
            "Make the text meaningfully shorter without removing important information."
        case .clearer:
            "Improve clarity and directness while preserving the meaning and factual claims."
        case .friendly:
            "Use a warm, friendly tone while preserving the meaning and factual claims."
        case .formal:
            "Use a polished, professional tone while preserving the meaning and factual claims."
        }
        return "\(base.trimmingCharacters(in: .whitespacesAndNewlines))\n\(transformation) Return only the replacement text."
    }
}

public struct ReviewableCorrectionProposal: Identifiable, Equatable, Sendable {
    public let id: UUID
    public let intent: BuddyRewriteIntent
    public let originalText: String
    public let proposedText: String
    public let change: BoundedTextChange

    public init(
        id: UUID = UUID(),
        intent: BuddyRewriteIntent,
        originalText: String,
        proposedText: String
    ) {
        self.id = id
        self.intent = intent
        self.originalText = originalText
        self.proposedText = proposedText
        self.change = BoundedTextChange(original: originalText, proposed: proposedText)
    }

    public var hasChanges: Bool { originalText != proposedText }
}

/// A single bounded changed span suitable for a compact keyboard diff. The
/// editor snapshot remains authoritative when a proposal is accepted.
public struct BoundedTextChange: Equatable, Sendable {
    public let commonPrefix: String
    public let originalChangedText: String
    public let proposedChangedText: String
    public let commonSuffix: String

    public init(original: String, proposed: String) {
        guard original != proposed else {
            commonPrefix = original
            originalChangedText = ""
            proposedChangedText = ""
            commonSuffix = ""
            return
        }

        let originalCharacters = Array(original)
        let proposedCharacters = Array(proposed)
        var prefixCount = 0
        while prefixCount < min(originalCharacters.count, proposedCharacters.count),
              originalCharacters[prefixCount] == proposedCharacters[prefixCount] {
            prefixCount += 1
        }

        var suffixCount = 0
        while suffixCount < originalCharacters.count - prefixCount,
              suffixCount < proposedCharacters.count - prefixCount,
              originalCharacters[originalCharacters.count - suffixCount - 1]
                == proposedCharacters[proposedCharacters.count - suffixCount - 1] {
            suffixCount += 1
        }

        commonPrefix = String(originalCharacters.prefix(prefixCount))
        originalChangedText = String(
            originalCharacters.dropFirst(prefixCount).dropLast(suffixCount)
        )
        proposedChangedText = String(
            proposedCharacters.dropFirst(prefixCount).dropLast(suffixCount)
        )
        commonSuffix = suffixCount == 0
            ? ""
            : String(originalCharacters.suffix(suffixCount))
    }
}

File: android/app/src/main/java/com/francescooddo/buddygrammar/core/ReviewableCorrectionProposal.kt
package com.francescooddo.buddygrammar.core

import java.util.UUID

enum class BuddyRewriteIntent(val title: String) {
    FIX("Fix"),
    SHORTEN("Shorten"),
    CLEARER("Clearer"),
    FRIENDLY("Friendly"),
    FORMAL("Formal");

    fun instruction(base: String): String {
        val transformation = when (this) {
            FIX -> "Fix only clear grammar, spelling, punctuation, and capitalization errors."
            SHORTEN -> "Make the text meaningfully shorter without removing important information."
            CLEARER -> "Improve clarity and directness while preserving the meaning and factual claims."
            FRIENDLY -> "Use a warm, friendly tone while preserving the meaning and factual claims."
            FORMAL -> "Use a polished, professional tone while preserving the meaning and factual claims."
        }
        return "${base.trim()}\n$transformation Return only the replacement text."
    }
}

data class ReviewableCorrectionProposal(
    val intent: BuddyRewriteIntent,
    val originalText: String,
    val proposedText: String,
    val id: UUID = UUID.randomUUID(),
) {
    val change = BoundedTextChange(originalText, proposedText)
    val hasChanges: Boolean get() = originalText != proposedText
}

data class BoundedTextChange(
    val commonPrefix: String,
    val originalChangedText: String,
    val proposedChangedText: String,
    val commonSuffix: String,
) {
    private constructor(values: Values) : this(
        values.commonPrefix,
        values.originalChangedText,
        values.proposedChangedText,
        values.commonSuffix,
    )

    constructor(original: String, proposed: String) : this(values(original, proposed))

    private data class Values(
        val commonPrefix: String,
        val originalChangedText: String,
        val proposedChangedText: String,
        val commonSuffix: String,
    )

    companion object {
        private fun values(original: String, proposed: String): Values {
            if (original == proposed) return Values(original, "", "", "")
            // Kotlin String indices address UTF-16 code units. Diffing those
            // directly can cut emoji, combining marks, or a ZWJ sequence into
            // invalid review spans. Work in user-perceived characters.
            val originalGraphemes = original.userPerceivedCharacters()
            val proposedGraphemes = proposed.userPerceivedCharacters()
            var prefix = 0
            while (
                prefix < minOf(originalGraphemes.size, proposedGraphemes.size) &&
                originalGraphemes[prefix] == proposedGraphemes[prefix]
            ) {
                prefix += 1
            }
            var suffix = 0
            while (
                suffix < originalGraphemes.size - prefix &&
                suffix < proposedGraphemes.size - prefix &&
                originalGraphemes[originalGraphemes.size - suffix - 1] ==
                proposedGraphemes[proposedGraphemes.size - suffix - 1]
            ) {
                suffix += 1
            }
            return Values(
                commonPrefix = originalGraphemes.subList(0, prefix).joinToString(""),
                originalChangedText = originalGraphemes
                    .subList(prefix, originalGraphemes.size - suffix)
                    .joinToString(""),
                proposedChangedText = proposedGraphemes
                    .subList(prefix, proposedGraphemes.size - suffix)
                    .joinToString(""),
                commonSuffix = originalGraphemes
                    .subList(originalGraphemes.size - suffix, originalGraphemes.size)
                    .joinToString(""),
            )
        }

    }
}

/** Android/JVM grapheme segmentation shared by diffing and bounded previews. */
internal fun String.userPerceivedCharacters(): List<String> {
    val boundaries = AndroidIcuGraphemeBoundaryProvider.validatedBoundaries(this)
    if (boundaries != null) {
        return boundaries.asSequence()
            .zipWithNext()
            .map { (start, end) -> substring(start, end) }
            .toList()
    }

    // Local JVM tests do not provide Android ICU. Java 17's \X implements
    // extended grapheme clusters, including emoji ZWJ and combining sequences.
    return runCatching {
        Regex("\\X").findAll(this).map { match -> match.value }.toList()
    }.getOrElse {
        // Last-resort validity fallback for an unusual regex runtime. It may
        // be less semantically rich, but never separates a surrogate pair.
        buildList {
            var start = 0
            while (start < length) {
                val end = start + Character.charCount(codePointAt(start))
                add(substring(start, end))
                start = end
            }
        }
    }
}
